File: 2024/01/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func calculateTotalDistance(leftNums, rightNums []int) int {
	left := make([]int, len(leftNums))
	right := make([]int, len(rightNums))
	copy(left, leftNums)
	copy(right, rightNums)

	sort.Ints(left)
	sort.Ints(right)

	totalDiff := 0
	for i := 0; i < len(left); i++ {
		if left[i] > right[i] {
			totalDiff += left[i] - right[i]
		} else {
			totalDiff += right[i] - left[i]
		}
	}
	return totalDiff
}

func calculateSimilarityScore(leftNums, rightNums []int) int {
	rightCounts := make(map[int]int)
	for _, num := range rightNums {
		rightCounts[num]++
	}

	totalScore := 0
	for _, leftNum := range leftNums {
		totalScore += leftNum * rightCounts[leftNum]
	}
	return totalScore
}

func main() {
	file, err := os.Open("input.csv")
	if err != nil {
		fmt.Println(err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var leftNums, rightNums []int

	for scanner.Scan() {
		numbers := strings.Fields(scanner.Text())
		if len(numbers) != 2 {
			continue
		}
		num1, _ := strconv.Atoi(numbers[0])
		num2, _ := strconv.Atoi(numbers[1])

		leftNums = append(leftNums, num1)
		rightNums = append(rightNums, num2)
	}

	totalDistance := calculateTotalDistance(leftNums, rightNums)
	similarityScore := calculateSimilarityScore(leftNums, rightNums)

	fmt.Printf("Total distance: %d\n", totalDistance)
	fmt.Printf("Similarity score: %d\n", similarityScore)
}

File: 2024/python/01/main.py
def calculate_total_distance(left_nums, right_nums):
    left = left_nums.copy()
    right = right_nums.copy()

    left.sort()
    right.sort()

    total_diff = sum(abs(l - r) for l, r in zip(left, right))
    return total_diff


def calculate_similarity_score(left_nums, right_nums):
    right_counts = {}
    for num in right_nums:
        right_counts[num] = right_counts.get(num, 0) + 1

    total_score = sum(num * right_counts.get(num, 0) for num in left_nums)
    return total_score


def main():
    left_nums = []
    right_nums = []

    try:
        with open("input") as file:
            for line in file:
                numbers = line.split()
                if len(numbers) != 2:
                    continue
                num1, num2 = map(int, numbers)
                left_nums.append(num1)
                right_nums.append(num2)

        total_distance = calculate_total_distance(left_nums, right_nums)
        similarity_score = calculate_similarity_score(left_nums, right_nums)

        print(f"Total distance: {total_distance}")
        print(f"Similarity score: {similarity_score}")

    except FileNotFoundError:
        print("Error: Could not open input file")


if __name__ == "__main__":
    main()
